src/inference/src/compiled_model.cpp

```diff
@@ -0,0 +1,131 @@
+                        "CompiledModel::input() must be called on a compiled model with exactly one parameter.");
+        return inputs.at(0);
+    });
+}
+
+const ov::Output<const ov::Node>& CompiledModel::input(size_t i) const {
+    OV_COMPILED_MODEL_CALL_STATEMENT({
+        OPENVINO_ASSERT(i < _impl->inputs().size(),
+                        "Cannot get input for index: ",
+                        i,
+                        " outputs size is ",
+                        _impl->inputs().size());
+        return _impl->inputs().at(i);
+    });
+}
+
+const ov::Output<const ov::Node>& CompiledModel::input(const std::string& tensor_name) const {
+    OV_COMPILED_MODEL_CALL_STATEMENT({
+        for (const auto& input : _impl->inputs()) {
+            if (input.get_names().count(tensor_name)) {
+                return input;
+            }
+        }
+        throw ov::Exception("Input for tensor name '" + tensor_name + "' is not found.");
+    });
+}
+
+const std::vector<ov::Output<const ov::Node>>& CompiledModel::outputs() const {
+    OV_COMPILED_MODEL_CALL_STATEMENT(return _impl->outputs());
+}
+
+const ov::Output<const ov::Node>& CompiledModel::output() const {
+    OV_COMPILED_MODEL_CALL_STATEMENT({
+        const auto& outputs = _impl->outputs();
+        OPENVINO_ASSERT(outputs.size() == 1,
+                        "CompiledModel::output() must be called on a compiled model with exactly one result.");
+        return outputs.at(0);
+    });
+}
+const ov::Output<const ov::Node>& CompiledModel::output(size_t i) const {
+    OV_COMPILED_MODEL_CALL_STATEMENT({
+        OPENVINO_ASSERT(i < _impl->outputs().size(),
+                        "Cannot get output for index: ",
+                        i,
+                        " outputs size is ",
+                        _impl->outputs().size());
+        return _impl->outputs().at(i);
+    });
+}
+const ov::Output<const ov::Node>& CompiledModel::output(const std::string& tensor_name) const {
+    OV_COMPILED_MODEL_CALL_STATEMENT({
+        for (const auto& output : _impl->outputs()) {
+            if (output.get_names().count(tensor_name)) {
+                return output;
+            }
+        }
+        throw ov::Exception("Output for tensor name '" + tensor_name + "' is not found.");
+    });
+}
+
+InferRequest CompiledModel::create_infer_request() {
+    OV_COMPILED_MODEL_CALL_STATEMENT(return {_impl->create_infer_request(), _so});
+}
+
+void CompiledModel::export_model(std::ostream& networkModel) {
+    OV_COMPILED_MODEL_CALL_STATEMENT(_impl->export_model(networkModel));
+}
+
+void CompiledModel::set_property(const AnyMap& config) {
+    OV_COMPILED_MODEL_CALL_STATEMENT(_impl->set_property(config));
+}
+
+Any CompiledModel::get_property(const std::string& name) const {
+    OV_COMPILED_MODEL_CALL_STATEMENT(return {_impl->get_property(name), {_so}});
+}
+
+RemoteContext CompiledModel::get_context() const {
+    OV_COMPILED_MODEL_CALL_STATEMENT(return {_impl->get_context()._impl, {_so}});
```

Comment:
is it possible that `get_context()` already provides an object with `so`? If yes, we loose / overwrite it here.